File: packages/ts/tests/e2e/sigint-interrupt.test.ts
// CLI e2e: SIGINT ends a run that is still in progress.
//
// Spawns the real launcher over a real config whose `on-file` hook blocks, so
// the core is provably mid-scan when the signal lands, then asserts the
// process dies of it. Nothing is mocked: real launcher, real process, real
// filesystem, real core.

import { spawn } from "node:child_process";
import { existsSync } from "node:fs";
import { mkdir, mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { setTimeout as delay } from "node:timers/promises";
import { fileURLToPath } from "node:url";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PKG_ROOT = join(__dirname, "..", "..");

const PKG: { bin: { dirsql: string } } = JSON.parse(
  await readFile(join(PKG_ROOT, "package.json"), "utf8"),
);
const CLI_ENTRY = join(PKG_ROOT, PKG.bin.dirsql);

async function waitUntilScanning(ready: string, timeoutMs = 30_000) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (existsSync(ready)) {
      return;
    }
    await delay(50);
  }
  throw new Error("the scan never reached the on-file hook");
}

describe("SIGINT during a scan (CLI)", () => {
  let dir: string;
  let configPath: string;
  let ready: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-sigint-e2e-"));
    configPath = join(dir, ".dirsql.toml");
    ready = join(dir, "scanning");
    await mkdir(join(dir, "data"), { recursive: true });
    await writeFile(join(dir, "data", "a.txt"), "hello");
    await writeFile(
      configPath,
      `
[[table]]
name = "files"
ddl = "CREATE TABLE files (path TEXT)"
glob = "data/*.txt"
on-file = "sh -c 'touch ${ready}; sleep 120; printf \\"[]\\"'"
`,
    );
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  it("kills the process", async () => {
    const proc = spawn(
      process.execPath,
      [CLI_ENTRY, "query", "SELECT * FROM files", "--config", configPath],
      { cwd: dir, detached: true, stdio: ["ignore", "ignore", "ignore"] },
    );
    const died = new Promise<[number | null, NodeJS.Signals | null]>(
      (resolve) => proc.on("close", (code, signal) => resolve([code, signal])),
    );

    try {
      await waitUntilScanning(ready);
      proc.kill("SIGINT");
      const settled = await Promise.race([
        died,
        delay(15_000).then(() => "still running" as const),
      ]);
      expect(settled).toEqual([null, "SIGINT"]);
    } finally {
      const pid = proc.pid;
      if (pid && proc.exitCode === null && proc.signalCode === null) {
        // The blocking hook is a child in the same group; take the group.
        process.kill(-pid, "SIGKILL");
        await died;
      }
    }
  }, 60_000);
});

File: packages/python/tests/e2e/sigint_interrupt_test.py
"""CLI e2e: `SIGINT` ends a run that is still in progress.

Spawns the real launcher over a real config whose `on-file` hook blocks, so
the core is provably mid-scan when the signal lands, then asserts the process
dies of it. No mocks: real process, real filesystem, real core.
"""

from __future__ import annotations

import os
import signal
import subprocess
import sys
import time

import pytest

_LAUNCH = "import sys; from dirsql.cli.main import main; sys.exit(main())"


def _wait_until_scanning(proc, ready, timeout=30.0):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if ready.exists():
            return
        assert proc.poll() is None, "the run exited before the scan began"
        time.sleep(0.05)
    raise AssertionError("the scan never reached the on-file hook")


def describe_sigint_during_a_scan():
    @pytest.fixture
    def scanning(tmp_path):
        root = tmp_path / "data"
        root.mkdir()
        (root / "a.txt").write_text("hello")
        ready = tmp_path / "scanning"
        cfg = root / ".dirsql.toml"
        cfg.write_text(
            "[[table]]\n"
            'name = "files"\n'
            'ddl = "CREATE TABLE files (path TEXT)"\n'
            'glob = "*.txt"\n'
            f'on-file = "sh -c \'touch {ready}; sleep 120; printf \\"[]\\"\'"\n'
        )

        proc = subprocess.Popen(
            [
                sys.executable,
                "-c",
                _LAUNCH,
                "--config",
                str(cfg),
                "SELECT * FROM files",
            ],
            cwd=str(root),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            start_new_session=True,
        )
        try:
            yield proc, ready
        finally:
            try:
                os.killpg(proc.pid, signal.SIGKILL)
            except OSError:
                pass
            proc.wait(timeout=10)

    def it_kills_the_process(scanning):
        proc, ready = scanning
        _wait_until_scanning(proc, ready)
        proc.send_signal(signal.SIGINT)
        assert proc.wait(timeout=15) == -signal.SIGINT
